src/main/java/org/tarantool/TarantoolClientImpl.java

```diff
@@ -63,7 +63,11 @@
         @Override
         public void run() {
             while (!Thread.currentThread().isInterrupted()) {
-                reconnect(0, thumbstone);
+                // Since LockSupport.park() can return at any time for no reason,
+                // we need to check if conditions are right to re-connect.
+                if (thumbstone != null) {
```

Comment:
`thumbstone` access is not guaranteed to be atomic, right? Maybe we should use alive.getCount() instead?

Reply:
Please correct me if I am wrong. We have three concurrent threads here.
The thread named `connector` reads `tombstone` variable and writes `null` only if it has read some non-null reference.
Other two threads (`reader` and `writer`) can only execute relevant code subsequently, since method `die()` is synchronized.
So, we have actually two concurrent threads.
The `reader`(`writer`) thread reads `tombstone` variable and writes non-null reference if it has read `null` previously.
So, if execution control will be transfered from one thread to another between read and write, 
nothing bad happens as the new thread will not perform the write.

The usage of `alive.getCount()` seems to be easier to reason about, though.

But, apparently it is not enough.

I noticed other potential issues that I'd like to bring your attention to.
Both `tombstone` and `alive` variables get updated by `die()` method long before the `unpark()` call and what is more important before the old threads/resources are cleaned-up. 
So, if connector thread will be awaken in that period (due to spurious return from `park()`) it 
may happen so that it will have time to reconnect and start new threads before the `die()` method is able to stop old threads. It possibly may attempt to stop newly spawned threads.

Probably we may move the initialization of 'alive' just before the unpark().
What do you think?

Also, the current code checks the connector thread state before calling `unpark()`.
This way, we may lose the opportunity to reconnect in case the thread state was
running and the next instruction was call to park(). I suggest we remove the check.

Reply:
read and writes are atomic for reference variables (esp. volatile) if you meant the possibility of word tearing by `atomic access`

Reply:
OK, I have removed the code changes from this PR.